kong/plugins/key-auth/handler.lua

```diff
@@ -112,6 +112,10 @@ local function do_authentication(conf)
     -- search the body, if we asked to
     if not v and conf.key_in_body then
       v = body[name]
+      -- duplicated keys in form-urlencoded requests
+      if not v and (body[name .. "[]"] or body[name .. "[1]"]) then
+        return nil, { status = 401, message = "Duplicate API key found" }
+      end
```

Comment:
I don't think that we should have such a branch in this plugin. We parse our `form-urlencoded` payloads with [ngx.req.get_post_args()](https://github.com/Kong/kong/blob/master/kong/pdk/request.lua#L625), and it doesn't support the notation we use to encode arrays; rightfully-so since it isn't standard anyway, and only used as a practical mean to mimic `application/json` payloads in the Admin API.

In my opinion, we should either:

1. Not add `[]` or `[1]` on leaf array nodes, which would preserve the previous encoding behavior and cause `ngx.req.get_post_args()` to return `{ "value1", "value2" }`, thus avoiding any changes in this plugin or in our test suite.
2. Disable the "duplicated API keys" test case for `form-urlencoded` payloads. Only run the test with `application/json` (it already is disabled for `multipart/form-data` content types.

Reply:
I have done something a bit different: I have left `key-auth` code code untouched, and instead added more tests which exactly document how things work now, without doing changes to key-auth:

* In json, if they key is duplicated, the error is "Duplicated key"
* In form-url-encoded:
  * `apikey=foo&apikey=foo` throws "Duplicated key"
  * `apikey[1]=foo&apikey[2]=foo` throws "Key not found"
  * `apikey[]=foo&apikey[]=foo` throws "Key not found"

I have not used automatic encoding and instead used direct strings for all the `form-url-encoded` tests. I think that makes things easier to follow.

Reply:
@kikito Got it. I reviewed the new `tests()` commit and it LGTM. I think this PR is good to go.

Reply:
@kikito Works too. Albeit our test client’s form encoder isn’t able to encode arrays that ngx_lua can decode anymore, which may cause us issues down the road (lack of test coverage when maintainers skip those tests for form-encoded payloads or even future changes to plugins similar to the one you originally made due to a lack of understanding around this new limitation).

Note that my above comment also implied that no changes were to be made to the plugin itself.